src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c

```diff
@@ -0,0 +1,222 @@
+#define SACIMEXGUS_HP(C)          ( SACIMEXGUS_CONTENT(C)->hp )
+#define SACIMEXGUS_FIRSTSTEP(C)   ( SACIMEXGUS_CONTENT(C)->firststep )
+
+/* ------------------
+ * Default parameters
+ * ------------------ */
+
+#define DEFAULT_K1E    RCONST(0.367)
+#define DEFAULT_K2E    RCONST(0.268)
+#define DEFAULT_K1I    RCONST(0.95)
+#define DEFAULT_K2I    RCONST(0.95)
+#define DEFAULT_BIAS   RCONST(1.5)
+#define TINY           RCONST(1.0e-10)
+
+
+/* -----------------------------------------------------------------
+ * exported functions
+ * ----------------------------------------------------------------- */
+
+/* -----------------------------------------------------------------
+ * Function to create a new ImExGus controller
+ */
+
+SUNAdaptController SUNAdaptController_ImExGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptControllerContent_ImExGus content;
+
+  /* Create an empty controller object */
+  C = NULL;
+  C = SUNAdaptController_NewEmpty(sunctx);
+  if (C == NULL) { return (NULL); }
+
+  /* Attach operations */
+  C->ops->gettype        = SUNAdaptController_GetType_ImExGus;
+  C->ops->estimatestep   = SUNAdaptController_EstimateStep_ImExGus;
+  C->ops->reset          = SUNAdaptController_Reset_ImExGus;
+  C->ops->setdefaults    = SUNAdaptController_SetDefaults_ImExGus;
+  C->ops->write          = SUNAdaptController_Write_ImExGus;
+  C->ops->seterrorbias   = SUNAdaptController_SetErrorBias_ImExGus;
+  C->ops->updateh        = SUNAdaptController_UpdateH_ImExGus;
+  C->ops->space          = SUNAdaptController_Space_ImExGus;
+
+  /* Create content */
+  content = NULL;
+  content = (SUNAdaptControllerContent_ImExGus)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    (void) SUNAdaptController_Destroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Fill content with default/reset values */
+  SUNAdaptController_SetDefaults_ImExGus(C);
+  SUNAdaptController_Reset_ImExGus(C);
+
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set ImExGus parameters
+ */
+
+int SUNAdaptController_SetParams_ImExGus(SUNAdaptController C,
+                                         sunrealtype k1e, sunrealtype k2e,
+                                         sunrealtype k1i, sunrealtype k2i)
+{
+  SACIMEXGUS_K1E(C) = k1e;
+  SACIMEXGUS_K2E(C) = k2e;
+  SACIMEXGUS_K1I(C) = k1i;
+  SACIMEXGUS_K2I(C) = k2i;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType_ImExGus(SUNAdaptController C)
+{ return SUN_ADAPTCONTROLLER_H; }
+
+int SUNAdaptController_EstimateStep_ImExGus(SUNAdaptController C, sunrealtype h,
+                                            int p, sunrealtype dsm, sunrealtype* hnew)
+{
```

Comment:
```suggestion
{
  if (C == NULL || hnew == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.